<a href="https://colab.research.google.com/github/Moogrob127/dataviz2025/blob/main/midterm2025.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ให้สืบหาฆาตรกร (10 คะแนน) และ ผู้บงการ (5 คะแนน) ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง Pandas City โดยเริ่มจากการตรวจสอบ ข้อมูลคดีในไฟล์ crime_scene_report.csv

# โค้ดช่วยในการการย้ายโฟลเดอร์

In [4]:
from google.colab import drive
import shutil
import os

# ✨ 1. Mount Google Drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [8]:
# ✨ 2. ตั้งค่าโฟลเดอร์ต้นทางและปลายทาง
source_folder = '/content/drive/MyDrive/Exam Data DataViz2025'  # โฟลเดอร์ที่ถูกแชร์ (คุณต้องเพิ่มทางลัดไว้ใน MyDrive แล้ว)
target_folder = '/content/drive/MyDrive/Exam_Data_DataViz2025'  # จะสร้างโฟลเดอร์

# ✨ 3. ตรวจสอบว่าปลายทางมีอยู่แล้วไหม
if not os.path.exists(target_folder):
    shutil.copytree(source_folder, target_folder)
    print(f"✅ คัดลอกสำเร็จ: {source_folder} ➡️ {target_folder}")
else:
    print(f"⚠️ โฟลเดอร์ {target_folder} มีอยู่แล้ว โปรดลบก่อนหรือเปลี่ยนชื่อปลายทาง")

✅ คัดลอกสำเร็จ: /content/drive/MyDrive/Exam Data DataViz2025 ➡️ /content/drive/MyDrive/Exam_Data_DataViz2025


# หาคดีฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง Pandas City

In [11]:
import pandas as pd

crime_report_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/crime_scene_report.csv')
crime_report_df

date     type                                        description  \
0     20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1     20180115   murder                  Life? Dont talk to me about life.   
2     20180115   murder  Mama, I killed a man, put a gun against his he...   
3     20180215   murder                         REDACTED REDACTED REDACTED   
4     20180215   murder  Someone killed the guard! He took an arrow to ...   
...        ...      ...                                                ...   
1223  20180430  bribery                                                 \n   
1224  20180430    fraud                  ‘Why not?’ said the March Hare.\n   
1225  20180430  assault                                                 \n   
1226  20180501  assault  be NO mistake about it: it was neither more no...   
1227  20180115   murder  Security footage shows that there were 2 witne...   

              city  
0              NYC  
1           Albany  
2             Reno  
3      Pandas City  
4      Pandas City  
...            ...  
1223  Garden Grove  
1224         Houma  
1225       Fontana  
1226       Trenton  
1227   Pandas City  

[1228 rows x 4 columns]

In [13]:
filtered_crime_df = crime_report_df[(crime_report_df['date'] == 20180115) & (crime_report_df['city'] == 'Pandas City')]
filtered_crime_df

date     type                                        description  \
11    20180115  assault  Hamilton: Lee, do you yield? Burr: You shot hi...   
12    20180115  assault                                   Report Not Found   
1227  20180115   murder  Security footage shows that there were 2 witne...   

             city  
11    Pandas City  
12    Pandas City  
1227  Pandas City

In [14]:
pd.set_option('display.max_colwidth', None)
display(filtered_crime_df)

date     type  \
11    20180115  assault   
12    20180115  assault   
1227  20180115   murder   

                                                                                                                                                                                    description  \
11                                                                                                                  Hamilton: Lee, do you yield? Burr: You shot him in the side! Yes he yields!   
12                                                                                                                                                                             Report Not Found   
1227  Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".   

             city  
11    Pandas City  
12    Pandas City  
1227  Pandas City

ภาพจากกล้องวงจรปิดแสดงให้เห็นว่ามีพยาน 2 คน พยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบนถนน "Northwestern Dr" ส่วนพยานคนที่สองชื่อแอนนาเบล อาศัยอยู่ที่ไหนสักแห่งบนถนน "Franklin Ave"

# หาพยาน

In [15]:
person_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/person.csv')
person_df.head()

id                 name  license_id  address_number address_street_name  \
0  10000   Christoper Peteuil      993845             624        Bankhall Ave   
1  10007  Kourtney Calderwood      861794            2791       Gustavus Blvd   
2  10010            Muoi Cary      385336             741     Northwestern Dr   
3  10016          Era Moselle      431897            1987       Wood Glade St   
4  10025         Trena Hornby      550890             276       Daws Hill Way   

         ssn  
0  747714076  
1  477972044  
2  828638512  
3  614621061  
4  223877684

## The first witness

 พยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบนถนน "Northwestern Dr"

In [21]:
first_witness = person_df[person_df['address_street_name'] == 'Northwestern Dr'].sort_values(by='address_number', ascending=False)
first_witness.head()

id            name  license_id  address_number address_street_name  \
499   14887  Morty Schapiro      118009            4919     Northwestern Dr   
811   17729  Lasonya Wildey      439686            3824     Northwestern Dr   
4886  53890  Sophie Tiberio      957671            3755     Northwestern Dr   
7040  73368  Torie Thalmann      773862            3697     Northwestern Dr   
9609  96595   Coretta Cubie      303645            3631     Northwestern Dr   

            ssn  
499   111564949  
811   917817122  
4886  442830147  
7040  341559436  
9609  378403829

## The second witness

ส่วนพยานคนที่สองชื่อแอนนาเบล อาศัยอยู่ที่ไหนสักแห่งบนถนน "Franklin Ave"

In [24]:
second_witness = person_df[(person_df['name'].str.contains('Annabel')) & (person_df['address_street_name'] == 'Franklin Ave')]
display(second_witness)

id            name  license_id  address_number address_street_name  \
665  16371  Annabel Miller      490173             103        Franklin Ave   

           ssn  
665  318771143

## คำให้การ

In [25]:
interview_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/interview.csv')
display(interview_df.head())

person_id  \
0      28508   
1      63713   
2      86208   
3      35267   
4      33856   

                                                                transcript  
0                                      ‘I deny it!’ said the March Hare.\n  
1                                                                       \n  
2                            way, and the whole party swam to the shore.\n  
3  lessons in here? Why, there’s hardly room for YOU, and no room at all\n  
4                                                                       \n

In [26]:
witness_interviews = interview_df[interview_df['person_id'].isin([14887, 16371])]
display(witness_interviews)

person_id  \
4988      14887   
4989      16371   

                                                                                                                                                                                                                           transcript  
4988  I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".  
4989                                                                                                            I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.

พยานคนแรกบอกว่า: ฉันได้ยินเสียงปืน แล้วก็เห็นผู้ชายคนหนึ่งวิ่งออกมา เขาถือกระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" มีแต่สมาชิกระดับ Gold เท่านั้นที่จะมีกระเป๋าแบบนั้น ชายคนนั้นขึ้นรถที่มีป้ายทะเบียน "H42W"

พยานคนที่สองบอกว่า: ฉันเห็นเหตุการณ์ฆาตกรรมเกิดขึ้น และฉันจำฆาตกรได้จากโรงยิมของฉันในขณะที่ฉันกำลังออกกำลังกายเมื่อวันที่ 9 มกราคมที่ผ่านมา

### แกะรอยฆาตกร จากคำให้การพยานคนที่ 1

กระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" มีแต่สมาชิกระดับ Gold

In [28]:
gym_members_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/get_fit_now_member.csv')
gym_members_df.head()

id  person_id                name  membership_start_date  \
0  NL318      65076     Everette Koepke               20170926   
1  AOE21      39426        Noe Locascio               20171005   
2  2PN28      63823  Jeromy Heitschmidt               20180215   
3  0YJ24      80651      Waneta Wellard               20171206   
4  3A08L      32858        Mei Bianchin               20170401   

  membership_status  
0              gold  
1           regular  
2            silver  
3              gold  
4            silver

In [29]:
filtered_members_by_id = gym_members_df[
    (gym_members_df['id'].astype(str).str.startswith('48Z')) &
    (gym_members_df['membership_status'] == 'gold')
]
display(filtered_members_by_id)

id  person_id           name  membership_start_date membership_status
181  48Z7A      28819   Joe Germuska               20160305              gold
182  48Z55      67318  Jeremy Bowers               20160101              gold

ป้ายทะเบียน "H42W"

In [30]:
drivers_license_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/drivers_license.csv')
display(drivers_license_df.head())

id  age  height eye_color hair_color  gender plate_number  car_make  \
0  100280   72      57     brown        red    male       P24L4U     Acura   
1  100460   63      72     brown      brown  female       XF02T6  Cadillac   
2  101029   62      74     green      green  female       VKY5KR     Scion   
3  101198   43      54     amber      brown  female       Y5NZ08    Nissan   
4  101255   18      79      blue       grey  female       5162Z1     Lexus   

  car_model  
0       MDX  
1       SRX  
2        xB  
3     Rogue  
4        GS

In [32]:
plate_H42W_df = drivers_license_df[drivers_license_df['plate_number'].str.contains('H42W')]
display(plate_H42W_df)

id  age  height eye_color hair_color  gender plate_number  \
915   183779   21      65      blue     blonde  female       H42W0X   
3529  423327   30      70     brown      brown    male       0H42W2   
6240  664760   21      71     black      black    male       4H42WR   

       car_make car_model  
915      Toyota     Prius  
3529  Chevrolet  Spark LS  
6240     Nissan    Altima

In [35]:
suspects_by_license = person_df[person_df['license_id'].isin([183779, 423327, 664760])]
display(suspects_by_license)

id            name  license_id  address_number  \
4664  51739  Tushar Chandra      664760             312   
6327  67318   Jeremy Bowers      423327             530   
7578  78193  Maxine Whitely      183779             110   

        address_street_name        ssn  
4664                 Phi St  137882671  
6327  Washington Pl, Apt 3A  871539279  
7578                Fisk Rd  137882671

ซึ่งจากข้อมูล จะเห็นได้ว่า คนที่มี กระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" มีแต่สมาชิกระดับ Gold และมีรถที่มีป้ายทะเบียนบางส่วน "H42W" คือคนที่มี id 67318 Jeremy Bowers

### แกะรอยฆาตกร จากคำให้การพยานคนที่ 2

ฉันจำฆาตกรได้จากโรงยิมของฉันในขณะที่ฉันกำลังออกกำลังกายเมื่อวันที่ 9 มกราคมที่ผ่านมา

In [33]:
check_in_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/get_fit_now_check_in.csv')
display(check_in_df.head())

membership_id  check_in_date  check_in_time  check_out_time
0         NL318       20180212            329             365
1         NL318       20170811            469             920
2         NL318       20180429            506             554
3         NL318       20180128            124             759
4         NL318       20171027            418            1019

In [34]:
january_9_check_ins = check_in_df[check_in_df['check_in_date'] == 20180109]
display(january_9_check_ins)

membership_id  check_in_date  check_in_time  check_out_time
113          X0643       20180109            957            1164
889          UK1F2       20180109            344             518
1329         XTE42       20180109            486            1124
1446         1AE2H       20180109            461             944
1703         6LSTG       20180109            399             515
2137         7MWHJ       20180109            273             885
2435         GE5Q8       20180109            367             959
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

In [45]:
gym_members_on_jan9 = gym_members_df[gym_members_df['id'].isin(january_9_check_ins['membership_id'])]
display(gym_members_on_jan9)

id  person_id               name  membership_start_date  \
9    X0643      15247     Shondra Ledlow               20170521   
70   UK1F2      28073   Zackary Cabotage               20170818   
96   XTE42      55662     Sarita Bartosh               20170524   
103  1AE2H      10815   Adriane Pelligra               20170816   
119  6LSTG      83186      Burton Grippe               20170214   
147  7MWHJ      31523  Blossom Crescenzo               20180309   
164  GE5Q8      92736      Carmen Dimick               20170618   
181  48Z7A      28819       Joe Germuska               20160305   
182  48Z55      67318      Jeremy Bowers               20160101   
183  90081      16371     Annabel Miller               20160208   

    membership_status  
9              silver  
70             silver  
96               gold  
103            silver  
119              gold  
147           regular  
164              gold  
181              gold  
182              gold  
183              gold

จะได้ผู้ต้องสงสัย ทั้งหมด 10 คน ดังตาราง

ซึ่งในนี้ก็มี Jeremy Bowers ที่เป็นผู้ต้องสงสัยในคำให้การของพยานคนที่ 1

### สรุปผู้ต้องสงสัย

จากเบาะแส

* กระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" มีแต่สมาชิกระดับ Gold
* ชายคนนั้นขึ้นรถที่มีป้ายทะเบียนบางส่วน "H42W"
* ฉันจำฆาตกรได้จากโรงยิมของฉันในขณะที่ฉันกำลังออกกำลังกายเมื่อวันที่ 9 มกราคมที่ผ่านมา

จะได้ตัวคนร้ายคือ Jeremy Bowers

# ผู้ต้องสงสัย

## คำให้การของ Jeremy Bowers

In [36]:
Jeremy_interviews = interview_df[interview_df['person_id'].isin([67318])]
display(Jeremy_interviews)

person_id  \
4990      67318   

                                                                                                                                                                                                                                              transcript  
4990  I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.\n

Jeremy Bowers ให้การว่า: ฉันถูกจ้างโดยผู้หญิงคนหนึ่งที่มีเงินมากมาย ฉันไม่รู้จักชื่อเธอ แต่ฉันรู้ว่าเธอสูงประมาณ 5 ฟุต 5 นิ้ว (65 นิ้ว) หรือ 5 ฟุต 7 นิ้ว (67 นิ้ว) เธอมีผมสีแดงและขับรถ Tesla Model S ฉันรู้ว่าเธอไปคอนเสิร์ต SQL Symphony 3 ครั้งในเดือนธันวาคม 2017

# ผู้บงการ

## ผู้บงการของ Jeremy Bowers

เธอสูงประมาณ 5 ฟุต 5 นิ้ว (65 นิ้ว) หรือ 5 ฟุต 7 นิ้ว (67 นิ้ว) เธอมีผมสีแดงและขับรถ Tesla Model S

In [37]:
mastermind_suspects = drivers_license_df[
    (drivers_license_df["height"].between(65, 67)) &
    (drivers_license_df["hair_color"] == "red") &
    (drivers_license_df["car_make"] == "Tesla") &
    (drivers_license_df["car_model"] == "Model S")
]
display(mastermind_suspects)

id  age  height eye_color hair_color  gender plate_number car_make  \
1105  202298   68      66     green        red  female       500123    Tesla   
2054  291182   65      66      blue        red  female       08CM64    Tesla   
9078  918773   48      65     black        red  female       917UU3    Tesla   

     car_model  
1105   Model S  
2054   Model S  
9078   Model S

In [43]:
# ดูว่าเป็นใครผ่าน โดยการนำ id ในตาราง drivers_license ไปเทียบกับ license_id ในตาราง person
suspects_by_license = person_df[person_df['license_id'].isin([202298, 291182, 918773])]
display(suspects_by_license)

id              name  license_id  address_number address_street_name  \
7664  78881          Red Korb      918773             107         Camerata Dr   
8977  90700     Regina George      291182             332           Maple Ave   
9985  99716  Miranda Priestly      202298            1883          Golden Ave   

            ssn  
7664  961388910  
8977  337169072  
9985  987756388

เธอไปคอนเสิร์ต SQL Symphony 3 ครั้งในเดือนธันวาคม 2017

In [38]:
facebook_checkins_df = pd.read_csv('/content/drive/MyDrive/Exam_Data_DataViz2025/facebook_event_checkin.csv')
display(facebook_checkins_df.head())

person_id  event_id  \
0      28508      5880   
1      63713      3865   
2      63713      3999   
3      63713      6436   
4      82998      4470   

                                                            event_name  \
0                      Nudists are people who wear one-button suits.\n   
1  but that's because it's the best book on anything for the layman.\n   
2                             If Murphy's Law can go wrong, it will.\n   
3     Old programmers never die.  They just branch to a new address.\n   
4                                 Help a swallow land at Capistrano.\n   

       date  
0  20170913  
1  20171009  
2  20170502  
3  20170926  
4  20171022

In [55]:
# กรองข้อมูล event_name = SQL Symphony Concert ในเดือนธันวาคม 2017
sql_symphony_checkins_december = facebook_checkins_df[
    (facebook_checkins_df["event_name"] == "SQL Symphony Concert") &
    (facebook_checkins_df["date"].astype(str).str.startswith('201712'))
]

# นับดูว่า person_id ไหนบ้างที่ซ้ำกัน 3 รอบ
sql_symphony_counts = sql_symphony_checkins_december['person_id'].value_counts()
mastermind_person_id = sql_symphony_counts[sql_symphony_counts == 3].index

# ดูว่า person_id ที่ซ้ำ 3 รอบ คนนั้นเป็นใคร โดยเทียบกับ id ในตาราง person_df
mastermind = person_df[person_df['id'].isin(mastermind_person_id)]
display(mastermind)

id              name  license_id  address_number address_street_name  \
1587  24556       Bryan Pardo      101191             703          Machine Ln   
9985  99716  Miranda Priestly      202298            1883          Golden Ave   

            ssn  
1587  816663882  
9985  987756388

ซึ่งคนที่ตรงกับเหตุการทั้งสองอย่าง คือ id 99716 Miranda Priestly

ผู้บงการของ Jeremy Bowers คือ Miranda Priestly

# สรุปคดี

**🕵️‍♂️ สรุปคดีฆาตกรรม Pandas City (15 มกราคม 2018)**

เหตุการณ์เกิดขึ้นวันที่ 15 มกราคม 2018 เมือง Pandas City มีการยิงปืน และพบผู้เสียชีวิต

**พยานคนแรก**

* ได้ยินเสียงปืนและเห็นชายคนหนึ่งวิ่งออกไป
* เขามีกระเป๋า Get Fit Now Gym (เฉพาะสมาชิก Gold)
* เลขสมาชิกขึ้นต้นด้วย 48Z
* หนีไปในรถที่มีป้ายทะเบียน H42W

นำไปสู่การตรวจสอบฐานข้อมูลยิม + ทะเบียนรถ → ตรงกับ Jeremy Bowers

**พยานคนที่สอง**

* ยืนยันว่า เห็นเหตุฆาตกรรม
* รู้จักฆาตกรจากฟิตเนส เพราะเห็นเขาเข้าไปออกกำลังกายเมื่อวันที่ 9 มกราคม 2018

นำไปสู่การตรวจสอบฐานข้อมูล → ตรงกับ Jeremy Bowers

**คำรับสารภาพ***

Jeremy Bowers ยอมรับว่าเขาเป็น ฆาตกร แต่ถูก ว่าจ้างโดยหญิงลึกลับที่ร่ำรวย

**การสืบสวนผู้บงการ**

* เบาะแส: ผู้หญิงสูง ~65–67 นิ้ว, ผมสีแดง, ขับ Tesla Model S
* เข้าร่วมงาน SQL Symphony Concert ถึง 3 ครั้งในเดือนธันวาคม 2017

ข้อมูลตรงกับ Miranda Priestly

**หาฆาตรกร คือ Jeremy Bowers** และ **ผู้บงการ คือ Miranda Priestly**